Add support to Dyson 360 Eye robot vacuum using new vacuum platform

CharlesBlonde opened the pull request:
## Description:

Add support to Dyson 360 Eye robot vacuum implementing the new `vacuum` platform added in PR #8623 

@azogue You did an amazing job with this platform! In less than 2 hours, I had a first fully working Dyson implementation. 

Nevertheless, I had to remove a test in the `vacuum` platform:

```python
if not config[DOMAIN]:
  return False
```
Because I'm calling the setup programaticly, without adding any
```yaml
vacuum:
```
I have to do that because I'm using the parent Dyson hub used also by fan/purifier. Do you see any issue by removing this test ?

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#3148

## Example entry for `configuration.yaml` (if applicable):

It's using the already existing Dyson hub configuration

```yaml
dyson:
  username: <email>
  password: <password>
  language: <language_code>
  devices:
    - device_id: <device_id> # eg: Dyson fan
      device_ip: <device_ip>
    - device_id: <device_id> # eg: Dyson robot vacuum
      device_ip: <device_ip>
```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

CharlesBlonde committed:
Add support to Dyson 360 Eye robot vacuum using new vacuum platform

CharlesBlonde committed:
Fix tests with Python 3.5

MartinHjelmare commented on homeassistant/components/vacuum/dyson.py:
```diff
@@ -0,0 +1,217 @@
+class Dyson360EyeDevice(VacuumDevice):
+    """Dyson 360 Eye robot vacuum device."""
+
+    def __init__(self, hass, device):
```
Don't pass in `hass`. It will be set on the entity when it has been added to home assistant.

MartinHjelmare commented on homeassistant/components/vacuum/dyson.py:
```diff
@@ -0,0 +1,217 @@
+            Dyson360EyeMode.FULL_CLEAN_FINISHED: "Finished",
+            Dyson360EyeMode.FULL_CLEAN_NEEDS_CHARGE: "Need charging"
+        }
+        if self._device.state.state in dyson_labels:
```
```py
return dyson_labels.get(self._device.state.state, self._device.state.state)
```

MartinHjelmare commented on homeassistant/components/dyson.py:
```diff
@@ -69,14 +69,17 @@ def setup(hass, config):
-                    _LOGGER.warning("Unable to connect to device %s",
-                                    dyson_device)
+                try:
+                    connected = dyson_device.connect(device["device_ip"])
```
What's the difference between this part of connection logic and the one at line 92-97? Would it be possible to make a connect function that is called wherever it is needed?

CharlesBlonde replied:
The first block is for manual connection (specifying device id/ip). The second one try to use mDNS protocol to discover fan/purifier devices (not working with robot vacuum yet). https://home-assistant.io/components/dyson/
It was the same method before but it was quite painful to use so I updated the underlying library and create 2 different methods.

MartinHjelmare reviewed (commented):
I assume there's no I/O going on in the entity properties? Three comments. I didn't look at the tests.

CharlesBlonde committed:
Code review

CharlesBlonde commented:
Thanks for your review.
I confirm there is no IO in the entity properties. Values are only updated with MQTT events and stored in `self._device` object.

MartinHjelmare reviewed (changes requested):
Fix the two import comments and I think this is good to go. :+1:

CharlesBlonde committed:
Code review - v2

CharlesBlonde commented:
I just updated the import orders (and also use an empty dict instead of none, it's better). Tell me if it's OK for you.

But before merging, I would like to have @azogue point of view because I updated the `vacuum` platform and I'm unable to check if it can break something. I don't think so but I can't be sure.

MartinHjelmare commented on homeassistant/components/vacuum/__init__.py:
```diff
@@ -168,9 +168,6 @@ def send_command(hass, command, params=None, entity_id=None):
 @asyncio.coroutine
 def async_setup(hass, config):
     """Set up the vacuum component."""
-    if not config[DOMAIN]:
```
This should not be here, so it's good to remove.

CharlesBlonde replied:
Ok, perfect.

azogue replied:
Yeah, I think I copied it from some other component, but I'm not sure why exactly, so better out...

MartinHjelmare commented on homeassistant/components/vacuum/dyson.py:
```diff
@@ -7,14 +7,14 @@
 import asyncio
 import logging
 
+from homeassistant.util.icon import icon_for_battery_level
```
No, please put this back below. Imports should be alphabetically sorted but following the greater rules of PEP8 for imports.

CharlesBlonde replied:
Ok, this time it should be fine. Next time I'll try to do it in the first commit :)

CharlesBlonde committed:
Code review - v3

azogue commented:
Hi @CharlesBlonde, congrats for this PR! 

It is great to see how the community works and how HA grows better and better each release!

MartinHjelmare commented:
@CharlesBlonde Please link to a docs PR asap.

azogue commented:
Yes, to include all three vacuum cleaners in the next release! (0.50.3 or 0.51.0)

... suggesting subtitle alias for the release: _The botvac revolution_ hehehe ...

CharlesBlonde commented:
I just submitted the PR for the documentation (https://github.com/home-assistant/home-assistant.github.io/pull/3148).
I added a comment on the `ha_category`

Ready for release 0.51.0 :)

Au8ust commented:
I try to use Dyson hub to add my 360 eye. But I am not sure which should be the device ID. I only see serial number in my settings page. I put it in the device ID but logs shows can't find the device. Could you please advice on?

MartinHjelmare commented:
Please use the forum or chat for support questions.

https://home-assistant.io/help/#communication-channels
